### ISSUE TITLE
Xamarin.Mac support

### ISSUE THREAD

**Issue by haritowa** [dsplaisted/PCLStorage#28]
Xamarin.Mac version doesn't work proper way(for example, LocalStorage returns ~/../Library, which incorrect by definition). [Here](https://github.com/xplatform/PCLStorage/tree/PCLStorage_XamarinMac) you can find fork, which repairs Mac compatibility

**Comment by dsplaisted** [comment]
I see that the fork creates a separate project which specifically targets Xamarin.Mac.  [As far as I know](https://twitter.com/dsplaisted/status/677662958030356480), it's not currently possible to build a project targeting Xamarin.Mac on Windows.

I think that libraries targeting .NET Framework 4.5 will probably work on Xamarin.Mac.  So what I'd like to try is to just compile the library for .NET Framework with [the code](https://github.com/xplatform/PCLStorage/blob/d7bef2c0d2e977319240bde8d0f33a6dc4d27d6f/src/PCLStorage.FileSystem.Desktop/DesktopFileSystem.cs#L31) that uses the correct path for LocalStorage, and see if it can be used from Xamarin.Mac.

Would you be willing to see if this works and send a PR for it?

**Comment by haritowa** [comment]
Sorry for delay, i was rly busy this weekend. As temporary solution, i use PCLStorage, PCLStorage.Abstractions & PCLStorage.FileSystem.Desktop source code in xamarin.mac project:
<img width="680" alt="screen shot 2015-12-21 at 03 04 08" src="https://cloud.githubusercontent.com/assets/6107211/11921034/ac623c96-a78f-11e5-8b33-b9a1b7c2d848.png">

with some changes(just remove #if derectives & use Filesystem || MAC || Portable code), for example

```
static IFileSystem CreateFileSystem()
{
    return new DesktopFileSystem();
}
```

My mac project use Xamarin.Mac Mobile as Target Framework, not Xamarin.Mac .Net 4.5

I can try to build everything u want on mac, if it'll help you :smile:

**Comment by Sparksteam** [comment]
I'm now getting this error with the LocalStorage on mac. Are there plans to implement a solution in the main branch and publishing an updated version? Thanks.

**Comment by Sparksteam** [comment]
So what do you think? Can we get this fixed soon?

Thanks

**Comment by haritowa** [comment]
@Sparksteam Still using temp solution ¯\_(ツ)_/¯

**Comment by Sparksteam** [comment]
@haritowa Hope not too much longer.

**Comment by haritowa** [comment]
@Sparksteam we have switched to native, so this doesn't make me suffer any more :joy: 

You can use source from this [fork](https://github.com/xplatform/PCLStorage/tree/PCLStorage_XamarinMac)

**Comment by Sparksteam** [comment]
Thanks @haritowa, I may have to use that fork eventually. Did you build a binding class for OS X, or are you going native across-the-board?

**Comment by haritowa** [comment]
@Sparksteam fully native both for iOS & OS X

**Comment by dsplaisted** [comment]
To be able to produce a package with Mac support, I need a way to be able to build a library that can run on Xamarin.Mac on a Windows machine. Possible ways of doing this include:

- Build a library which targets .NET Framework 4.5, and rely on the fact that it should "just work" when running on Xamarin.Mac
- Build against the Xamarin.Mac reference assemblies.  Xamarin would need to release them in a publicly consumable form (preferrably a NuGet package)
- Target the [.NET Platform Standard](https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-platform-standard.md) v1.3.  Xamarin.Mac would need to support .NETStandard v1.3, which I don't think it does yet.

**Comment by Sparksteam** [comment]
Why would it need to run on Windows? Is there bait-and-switch going on in the NuGet?

**Comment by dsplaisted** [comment]
@Sparksteam It doesn't need to run on Windows, but I need to be able to build it on Windows.  I don't own a Mac, so I can't use one to build the library for Xamarin.Mac.  Even if I did, I would strongly prefer to enable everything to be built on a single machine, instead of having to build some of the files on Windows, some on a Mac, and then copying them together and creating a NuGet package out of them.

**Comment by dsplaisted** [comment]
@Sparksteam And yes, PCL Storage is a canonical example of bait-and-switch.  The NuGet package contains a portable "reference assembly" that PCLs can compile against, and then implementations compiled for each platform supported.

**Comment by Sparksteam** [comment]
Yeah Daniel, it is somewhat cumbersome. And expensive to buy a MAC. I’d be more than happy to build binaries for you and upload them somewhere.

 

Allan

**Comment by Sparksteam** [comment]
@haritowa - Is this only change you had to make to the master code, in DesktopFileSystem.LocalStorage?

#elif MAC
   var documents = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
   var localAppData = Path.Combine(documents, "Library");
#else

and adding the necessary MAC preprocessor symbol, of course.

Thanks

**Comment by haritowa** [comment]
@Sparksteam unfortunately, i can't remember all details, i just remove all #IF sections for other platforms :)

**Comment by haritowa** [comment]
[PCLStorage.Mac.zip](https://github.com/dsplaisted/PCLStorage/files/180767/PCLStorage.Mac.zip)
@Sparksteam Try this

**Comment by Sparksteam** [comment]
Thanks @haritowa . I diff'd the files with the current master and although there were some differences, it doesn't look like any of them contain any MAC specific changes. The one change I referenced last week in your DesktopFileSystem.LocalStorage might be all that's required. I will give that a try and let you know.

**Comment by Sparksteam** [comment]
@haritowa - I worked around this problem in my code by simply implementing my own GetStorageFolder with this code;

IFolder storageFolder;

#if _MAC
   var documents = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
					var localAppData = documents + "/Library";
   storageFolder = await FileSystem.Current.GetFolderFromPathAsync(localAppData);
#else
   storageFolder = FileSystem.Current.LocalStorage;
#endif

@dsplaisted  - My offer still stands if you'd like me to build some binaries for you. The fix seems simple enough, but you will need that MAC project.

Thanks guys.

**Comment by dsplaisted** [comment]
@Sparksteam Can you send a PR to add support for building on the Mac?  Once we have that merged then you can build the Mac DLL and send it to me.

Thanks,
Daniel

**Comment by Sparksteam** [comment]
Ok Daniel - PR #38 with the changes I think are required. I'm not sure how to go about testing it though. I have all the release binaries for PCLStorage.OSX. Let me know how you would like me to get them to you. If you build and publish a pre-release NuGet package with these changes I would be more than happy to test it for you.

Thanks,

Allan

**Comment by soroshsabz** [comment]
ITNOA

Is there any update?

**Comment by martinobordin** [comment]
Is it working?
